File: PartitionArrayMaxSum.cpp
class Solution {
public:
    int dp[510];
    int solve(int i, int k, vector<int>& arr) {
        if (i == arr.size())
            return 0;
        if (dp[i] != -1)
            return dp[i];
        int ans = 0;
        int mx = 0;
        for (int j = i; j < i + k && j < arr.size(); j++) {
            mx = max(mx, arr[j]);
            ans = max(ans, mx * (j - i + 1) + solve(j + 1, k, arr));
        }
        return dp[i] = ans;
    }

    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        int n = arr.size();
        memset(dp, -1, sizeof(dp));
        return solve(0, k, arr);
    }
};

File: PartitionArrayMaxSum.go
var dp [510]int

func solve(i, k int, arr []int) int {
	if i == len(arr) {
		return 0
	}
	if dp[i] != -1 {
		return dp[i]
	}
	ans := 0
	mx := 0
	for j := i; j < i+k && j < len(arr); j++ {
		mx = int(math.Max(float64(mx), float64(arr[j])))
		ans = int(math.Max(float64(ans), float64(mx)*(float64(j-i+1))+float64(solve(j+1, k, arr))))
	}
	dp[i] = ans
	return ans
}

func maxSumAfterPartitioning(arr []int, k int) int {
	for i := range dp {
		dp[i] = -1
	}
	return solve(0, k, arr)
}
